### DISCUSSION TITLE
LFX Mentorship 2022 03-Sept-Nov Challenge - For #1741

### DISCUSSION THREAD

**Discussion by hydai** [WasmEdge/WasmEdge#1751] [Show and tell]
### Introduction

For those who want to apply for LFX mentorship for #1741, this is a pretest for the application.
This issue is about how to serialize and deserialize WasmEdge internal data structures. However, using a real WasmEdge data structure may be too complex in this challenge. To simplify the challenge, please use the following basic C++ class instead:

```cpp
class Data {
  int32_t Int32;
  int64_t Int64;
  float Float;
  double Double;
  std::vector<uint8_t> U8Array;
};
```

The signature of the `serialize` function should at least have the following information, but you can append any useful parameters or richer return values if you want:

```cpp
std::vector<uint8_t> serialize(Data &D);
```

The signature of the `deserialize` function should at least have the following information, but you can append any useful parameters or richer return values if you want:

```cpp
Data deserialize(std::vector<uint8_t> &Bytes);
```

### Requirements

1. Create a repository. You should define and implement your own `serialize` and `deserialize` functions in the repo.
2. All of the integer types should use **LEB128** for encoding and decoding.
3. After you finish the implementation, you should write some tests to verify your codes.
4. Before you submit your repo, please make sure there is a README file, which describes how to build and execute, how you design the serialized format, and how you test and verify your implementation.
5. Post your repo's GitHub link as a comment under this discussion.

### Deadline

Please refer to [the timeline of LFX mentorship 2022 03-Sept-Nov](https://github.com/cncf/mentoring/tree/main/lfx-mentorship/2022/03-Sept-Nov#timeline).
The final date of valid submissions will be August 31, 5:00 PM PDT

### Reference
1. LEB128: https://en.wikipedia.org/wiki/LEB128

Good luck!

**Comment by OmkarAcharekar** [comment]
Hello @hydai,  Sir I finished the challenge.
Github Repo : [Link](https://github.com/OmkarAcharekar/Serialize_and_Deserialize_DataStructure)

If there is any improvement that I  could make or corrections/ mistakes.  Please let me know. It was fun and interesting learning and implementing it.

Thank you,
